specs/01-sharded-archiving.md

```diff
@@ -0,0 +1,247 @@
+
+This section collects a list of data structures required for the operation of the protocol:
+
+- `segment_commitments[]`: hash map of KZG commitments computed over pieces in a segment and stored
+  in the runtime of each shard. This data structure represents the segment history of the shard and
+  include both, local segment commitments, and segment commitments from child shards (through
+  `SuperSegment`s). Each of them should be accordingly tagged as such.
+- `shard_blocks[]`: hash map of blocks committed from child shards. It stores the history of the
+  shard in the parent, until these blocks are made available in a new segment.
+- `recorded_history_buffer`: Buffer storing segments (i.e. the local history) of a local shard
+
+### Preparing the history of a shard
+
+#### 1. Genesis Block Archiving
+
+- Input: Genesis block of the shard.
+- The genesis block is archived as soon as it is produced. We extend the encoding of the genesis
+  block with extra pseudorandom data up to `RECORDED_HISTORY_SEGMENT_SIZE`, such that the very first
+  archived segment can be produced right away, bootstrapping the farming process. This extra data is
+  added to the end of the SCALE-encoded block, so during decoding of the genesis block it'll be
+  ignored.
+- This becomes the first segment of the shard's history.
+- Output: First archived segment.
+
+#### 2. Block History Buffering
+
+- Input: Blocks produced in the shard.
+- Once any block produced after genesis becomes `CONFIRMATION_DEPTH_K`-deep, it is included in the
+  recorded history. New blocks are added to recorded history that is a buffer of capacity
+  `RECORDED_HISTORY_SEGMENT_SIZE`.
+- When added to the buffer, blocks are turned into `SegmentItem`s.
+- Each segment will have the parent `segment_header` included as the first item. Each
+  `segment_header` includes the hash of the previous segment header and the segment commitment of
+  the previous segment. Together, segment headers form a chain that is used for quick and efficient
+  verification of pieces corresponding to the actual archival history of the blockchain.
+- Segment items are combined into segments. Each segment contains at least two of the following
+  `SegmentItem`s, in this order:
+- The previous segment’s SegmentHeader
+  - BlockContinuation, if remained from the previous segment
+  - Block(s), as many as fit fully into the current segment, may be none
+  - BlockStart, if the next block doesn't fit within the current segment
+  - Padding (zero or more) in case padding bytes are necessary to complete the segment to
+    `RECORDED_HISTORY_SEGMENT_SIZE`
+- Output: Buffered SegmentItems.
+
+#### 3. Block Info Construction
+
+- Input: Blocks produced in the shard.
+- When a new block is produced and included to the chain in a shard, farmers (and any other node in
+  the system), extend the block using an error correction scheme to generate parity shares:
+  1. The block is split into `M = ceil(block_size/l)` shares of `l` bytes.
```

Comment:
I would prefer to not use single-letter and otherwise cryptic variables/parameters/constants. For example this is much more readable to me:
```suggestion
  1. The block is split into `num_shares = ceil(block_size / SHARE_SIZE)` shares of `SHARE_SIZE` bytes.
```

`SHARE_SIZE` is uppercase because it is a constant, the rest are runtime variables.

It is a lot more verbose, but requires much less brain power to process and it is easier to search other uses of the same variable (like `num_shares`) in specs and code. Something like `l` will have huge number of matches in basically every file, making it very painful to find where it is used.